File: beecrowd/Estruturas e bibliotecas/ 2018_OlimpíadasDeNatal.java
import java.util.*;

class Pais{
    String nome;
    int ouro,prata,bronze;

    Pais(String nome,int o,int p,int b){
        this.nome=nome;
        this.ouro=o;
        this.prata=p;
        this.bronze=b;
    }
}

public class Main {
public static boolean buscarPais(String nome,Pais[] arr,int tam,int t){
    for(int i=0;i<tam;i++){
        if(arr[i].nome.compareTo(nome)==0){
            if(t==1)arr[i].ouro+=1;
            if(t==2)arr[i].prata+=1;
            if(t==3)arr[i].bronze+=1;
            return true;
        }
    }

    return false;
}

public static boolean cmp(Pais i,Pais j){
    if(i.ouro!=j.ouro)return i.ouro>j.ouro;
    
    //empate em ouro
    if(i.prata!=j.prata)return i.prata>j.prata;
    
    //empate em prata
    if(i.bronze!=j.bronze)return i.bronze>j.bronze;

    //empate nas 3
    return i.nome.compareTo(j.nome)<0;
}

public static void insertionSort(Pais[] arr,int tam){
    for(int i=1;i<tam;i++){
        Pais key=arr[i];
        int j=i-1;
        while(j>=0 && cmp(key,arr[j])){
            arr[j+1]=arr[j];
            j--;
        }
        arr[j+1]=key;
    }
}

public static void main(String[] args){
        Scanner sc=new Scanner(System.in); 
        
        Pais[] arr=new Pais[301];
        int tam=0;
        while(sc.hasNextLine()){
            String esporte=sc.nextLine();
          
            String primeiro=sc.nextLine();
            if(!buscarPais(primeiro,arr,tam,1)){  //1=ouro, 2=prata, 3=bronze
                arr[tam++]=new Pais(primeiro,1,0,0);
            }
            
            String segundo=sc.nextLine();
            if(!buscarPais(segundo,arr,tam,2)){
                arr[tam++]=new Pais(segundo,0,1,0);
            }

            String terceiro=sc.nextLine();
            if(!buscarPais(terceiro,arr,tam,3)){
                arr[tam++]=new Pais(terceiro,0,0,1);
            }
        }

        insertionSort(arr,tam);
        System.out.println("Quadro de Medalhas");

        for(int i=0;i<tam;i++){
            System.out.printf("%s %d %d %d\n",arr[i].nome,arr[i].ouro,arr[i].prata,arr[i].bronze);
        }

        sc.close();
    }
}

File: beecrowd/Estruturas e bibliotecas/2018_OlimpíadasDeNatal.c
#include <stdio.h>
#include <string.h>

typedef struct Pais{
    char nome[50];
    int o,p,b;
}Pais;

int buscar(char *s,Pais arr[],int n){
    for(int i=0;i<n;i++){
        if(strcmp(arr[i].nome,s)==0)return i;
    }
    return -1;
}

int cmp(Pais i,Pais j){
    if(i.o!=j.o)return i.o>j.o;

    if(i.p!=j.p)return i.p>j.p;

    if(i.b!=j.b)return i.b>j.b;
    
    if(strcmp(i.nome,j.nome)<0)return 1;

    return 0;
}

int insertionSort(Pais arr[],int n){
    for(int i=1;i<n;i++){
        Pais key=arr[i];
        int j=i-1;

        while(j>=0 && cmp(key,arr[j])){
            arr[j+1]=arr[j];
            j--;
        }
        arr[j+1]=key;
    }
}

int main(){
    char modalidade[150];

    Pais arr[350];
    int indx=0;
    
    //int n=5;

    while(fgets(modalidade,50,stdin)){
        for(int i=0;i<3;i++){
            char p[50];
            fgets(p, 50,stdin);
            
            int aux=buscar(p,arr,indx);
            if(aux==-1){ //pais não existe
                Pais a;
                strcpy(a.nome,p);
                if(i==0){
                    a.o=1;
                    a.p=0;
                    a.b=0;
                }else if(i==1){
                    a.o=0;
                    a.p=1;
                    a.b=0;
                }else{
                    a.o=0;
                    a.p=0;
                    a.b=1;
                }
                arr[indx++]=a;
            }else{
                if(i==0)arr[aux].o+=1;
                if(i==1)arr[aux].p+=1;
                if(i==2)arr[aux].b+=1;
            }
        }
    }
    
    insertionSort(arr,indx);

    printf("Quadro de Medalhas\n");
    for(int i=0;i<indx;i++){
        arr[i].nome[strlen(arr[i].nome)-1]='\0';
        printf("%s %d %d %d\n",arr[i].nome,arr[i].o,arr[i].p,arr[i].b);
    }
    
    return 0;
}
